// @neongd/neo-dapi/dist/esm/errors.js
import { getStandardErrorResponse, StandardErrorCodes } from '@neongd/json-rpc';
export var DapiErrorCodes;
(function (DapiErrorCodes) {
    DapiErrorCodes[DapiErrorCodes["UserRejected"] = -32001] = "UserRejected";
    DapiErrorCodes[DapiErrorCodes["UnsupportedNetwork"] = -32002] = "UnsupportedNetwork";
    DapiErrorCodes[DapiErrorCodes["NoAccount"] = -32003] = "NoAccount";
    DapiErrorCodes[DapiErrorCodes["InsufficientFunds"] = -32004] = "InsufficientFunds";
    DapiErrorCodes[DapiErrorCodes["RemoteRpcError"] = -32005] = "RemoteRpcError";
})(DapiErrorCodes || (DapiErrorCodes = {}));
const ERROR_MESSAGE_MAP = {
    [DapiErrorCodes.UserRejected]: 'User rejected',
    [DapiErrorCodes.UnsupportedNetwork]: 'Unsupported network',
    [DapiErrorCodes.NoAccount]: 'No account',
    [DapiErrorCodes.InsufficientFunds]: 'Insufficient funds',
    [DapiErrorCodes.RemoteRpcError]: 'Remote rpc error',
};
export function isDapiErrorCode(code) {
    return Object.values(DapiErrorCodes).includes(code);
}
export function getDapiErrorResponse(code) {
    if (isDapiErrorCode(code)) {
        return {
            code,
            message: ERROR_MESSAGE_MAP[code],
        };
    }
    return getStandardErrorResponse(StandardErrorCodes.InternalError);
}
