// @neongd/json-rpc/dist/cjs/formatters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatErrorResponse = exports.formatJsonRpcError = exports.formatJsonRpcResult = exports.formatJsonRpcNotification = exports.formatJsonRpcRequest = exports.payloadId = void 0;
const errors_1 = require("./errors");
let lastPayloadId = 0;
function payloadId() {
    return ++lastPayloadId;
}
exports.payloadId = payloadId;
function formatJsonRpcRequest(method, params, id) {
    return {
        id: id !== null && id !== void 0 ? id : payloadId(),
        jsonrpc: '2.0',
        method,
        params,
    };
}
exports.formatJsonRpcRequest = formatJsonRpcRequest;
function formatJsonRpcNotification(method, params) {
    return {
        jsonrpc: '2.0',
        method,
        params,
    };
}
exports.formatJsonRpcNotification = formatJsonRpcNotification;
function formatJsonRpcResult(id, result) {
    return {
        id,
        jsonrpc: '2.0',
        result,
    };
}
exports.formatJsonRpcResult = formatJsonRpcResult;
function formatJsonRpcError(id, error) {
    return {
        id,
        jsonrpc: '2.0',
        error: formatErrorResponse(error),
    };
}
exports.formatJsonRpcError = formatJsonRpcError;
function formatErrorResponse(error) {
    var _a;
    let code = errors_1.StandardErrorCodes.InternalError;
    let message = errors_1.getStandardErrorResponse(code).message;
    let data;
    if (error == null) {
        // noop
    }
    else if (typeof error === 'string') {
        message = error;
    }
    else if (error instanceof Error) {
        code = errors_1.isValidErrorCode(error.code) ? error.code : code;
        message = error.message;
        data = error.data;
    }
    else {
        code = errors_1.isValidErrorCode(error.code) ? error.code : code;
        message = (_a = error.message) !== null && _a !== void 0 ? _a : errors_1.getStandardErrorResponse(code).message;
        data = error.data;
    }
    return { code, message, data };
}
exports.formatErrorResponse = formatErrorResponse;
