// @neongd/json-rpc/dist/cjs/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isJsonRpcError = exports.isJsonRpcResult = exports.isJsonRpcResponse = exports.isJsonRpcNotification = exports.isJsonRpcRequest = exports.isJsonRpcPayload = void 0;
function isJsonRpcPayload(payload) {
    return 'jsonrpc' in payload && payload.jsonrpc === '2.0';
}
exports.isJsonRpcPayload = isJsonRpcPayload;
function isJsonRpcRequest(payload) {
    return isJsonRpcPayload(payload) && 'id' in payload && 'method' in payload;
}
exports.isJsonRpcRequest = isJsonRpcRequest;
function isJsonRpcNotification(payload) {
    return isJsonRpcPayload(payload) && !('id' in payload) && 'method' in payload;
}
exports.isJsonRpcNotification = isJsonRpcNotification;
function isJsonRpcResponse(payload) {
    return isJsonRpcPayload(payload) && (isJsonRpcResult(payload) || isJsonRpcError(payload));
}
exports.isJsonRpcResponse = isJsonRpcResponse;
function isJsonRpcResult(payload) {
    return 'result' in payload;
}
exports.isJsonRpcResult = isJsonRpcResult;
function isJsonRpcError(payload) {
    return 'error' in payload;
}
exports.isJsonRpcError = isJsonRpcError;
