// @rentfuse-labs/neo-wallet-adapter-o3/lib/adapter.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { BaseWalletAdapter, WalletAccountError, WalletDisconnectionError, } from '@rentfuse-labs/neo-wallet-adapter-base';
import neo3Dapi from 'neo3-dapi';
const DEFAULT_WALLET_CONFIG = { options: null };
// Reference at https://neo3dapidocs.o3.network/#getting-started (Taken on 10/11/21)
export class O3WalletAdapter extends BaseWalletAdapter {
    constructor(config = DEFAULT_WALLET_CONFIG) {
        super();
        // Arrow function to bind this correctly in event listener
        this._disconnected = () => {
            this.disconnect();
        };
        this._address = null;
        this._connecting = false;
        this._options = config.options;
    }
    get address() {
        return this._address;
    }
    get ready() {
        return typeof window !== 'undefined';
    }
    get connecting() {
        return this._connecting;
    }
    get connected() {
        return !!this._address;
    }
    connect() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (this.connected || this.connecting)
                    return;
                this._connecting = true;
                // Taken from o3 specs
                let account;
                try {
                    // O3 asks the user to connect the dapp when calling the getAccount method
                    account = yield neo3Dapi.getAccount();
                }
                catch (error) {
                    throw new WalletAccountError(error === null || error === void 0 ? void 0 : error.message, error);
                }
                if (!account)
                    throw new WalletAccountError();
                this._address = account.address;
                // Add a listener to cleanup of disconnection
                neo3Dapi.addEventListener(neo3Dapi.Constants.EventName.DISCONNECTED, this._disconnected);
                neo3Dapi.addEventListener(neo3Dapi.Constants.EventName.ACCOUNT_CHANGED, this._disconnected);
                neo3Dapi.addEventListener(neo3Dapi.Constants.EventName.NETWORK_CHANGED, this._disconnected);
                this.emit('connect');
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
            finally {
                this._connecting = false;
            }
        });
    }
    disconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                neo3Dapi.removeEventListener(neo3Dapi.Constants.EventName.DISCONNECTED);
                neo3Dapi.removeEventListener(neo3Dapi.Constants.EventName.ACCOUNT_CHANGED);
                neo3Dapi.removeEventListener(neo3Dapi.Constants.EventName.NETWORK_CHANGED);
                // TODO: How?
                //await neo3Dapi.disconnect();
                this._address = null;
            }
            catch (error) {
                this.emit('error', new WalletDisconnectionError(error === null || error === void 0 ? void 0 : error.message, error));
            }
            this.emit('disconnect');
        });
    }
    invokeRead(request) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield neo3Dapi.invokeRead({
                    scriptHash: request.scriptHash,
                    operation: request.operation,
                    args: request.args,
                    signers: request.signers,
                });
                return this._responseToReadResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    invokeReadMulti(request) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield neo3Dapi.invokeReadMulti({
                    invokeReadArgs: request.invocations,
                    signers: request.signers,
                });
                return this._responseToReadResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    invoke(request) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield neo3Dapi.invoke({
                    scriptHash: request.scriptHash,
                    operation: request.operation,
                    args: request.args,
                    signers: request.signers,
                    fee: request.fee,
                    extraSystemFee: request.extraSystemFee,
                    broadcastOverride: request.broadcastOverride,
                });
                return this._responseToWriteResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    invokeMulti(request) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield neo3Dapi.invokeMulti({
                    invokeArgs: request.invocations,
                    signers: request.signers,
                    fee: request.fee,
                    extraSystemFee: request.extraSystemFee,
                    broadcastOverride: request.broadcastOverride,
                });
                return this._responseToWriteResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    getNetworks() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield neo3Dapi.getNetworks();
                return this._responseToGetNetworksResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    signMessage(request) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield neo3Dapi.signMessage({
                    message: request.message,
                });
                return this._responseToSignMessageResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    _responseToReadResult(response) {
        // If the state is halt it means that everything went well
        if (response.state === 'HALT') {
            return {
                status: 'success',
                data: {
                    script: response.script,
                    state: response.state,
                    gasConsumed: response.gas_consumed,
                    exception: null,
                    stack: response.stack,
                },
            };
        }
        // Otherwise an error occurred and so return it
        // TODO: message and code where?
        return {
            status: 'error',
            message: undefined,
            code: undefined,
        };
    }
    _responseToWriteResult(response) {
        return {
            status: 'success',
            data: {
                txId: response.txid,
            },
        };
    }
    _responseToGetNetworksResult(response) {
        return {
            status: 'success',
            data: {
                networks: response.networks,
                defaultNetwork: response.defaultNetwork,
            },
        };
    }
    _responseToSignMessageResult(response) {
        return {
            status: 'success',
            data: {
                publicKey: response.publicKey,
                data: response.data,
                salt: response.salt,
                message: response.message,
            },
        };
    }
}
