// @rentfuse-labs/neo-wallet-adapter-neoline/lib/adapter.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { BaseWalletAdapter, pollUntilReady, WalletAccountError, WalletConnectionError, WalletDisconnectionError, WalletError, WalletNotConnectedError, } from '@rentfuse-labs/neo-wallet-adapter-base';
import { NeoLineInit, NeoLineN3Init, } from './utils/neoline';
const DEFAULT_WALLET_CONFIG = { options: null };
// Reference at https://neoline.io/dapi/N3.html (Taken on 08/11/21)
export class NeoLineWalletAdapter extends BaseWalletAdapter {
    constructor(config = DEFAULT_WALLET_CONFIG) {
        super();
        // Arrow function to bind this correctly in event listener
        this._disconnected = () => {
            this.disconnect();
        };
        this._address = null;
        this._connecting = false;
        this._options = config.options;
        if (!this.ready)
            pollUntilReady(this, config.pollInterval || 1000, config.pollCount || 3);
    }
    get address() {
        return this._address;
    }
    get ready() {
        return typeof window !== 'undefined' && window.NEOLineN3 !== 'undefined';
    }
    get connecting() {
        return this._connecting;
    }
    get connected() {
        return !!this._address;
    }
    connect() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (this.connected || this.connecting)
                    return;
                this._connecting = true;
                try {
                    // Get the neoline client initializing the wallet
                    this._client = yield NeoLineN3Init();
                    this._clientCommon = yield NeoLineInit();
                }
                catch (error) {
                    if (error instanceof WalletError)
                        throw error;
                    throw new WalletConnectionError(error === null || error === void 0 ? void 0 : error.message, error);
                }
                if (!this._client || !this._clientCommon)
                    throw new WalletAccountError();
                let account;
                try {
                    // NeoLine asks the user to connect the dapp when calling the getAccount method
                    account = yield this._client.getAccount();
                }
                catch (error) {
                    throw new WalletAccountError(error === null || error === void 0 ? void 0 : error.message, error);
                }
                if (!account)
                    throw new WalletAccountError();
                this._address = account.address;
                // Add a listener to cleanup on changes condition
                window.addEventListener('NEOLine.NEO.EVENT.DISCONNECTED', this._disconnected);
                window.addEventListener('NEOLine.NEO.EVENT.ACCOUNT_CHANGED', this._disconnected);
                window.addEventListener('NEOLine.NEO.EVENT.NETWORK_CHANGED', this._disconnected);
                this.emit('connect');
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
            finally {
                this._connecting = false;
            }
        });
    }
    disconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            const client = this._client;
            if (client) {
                try {
                    // TODO: How?
                    //await this._client.disconnect();
                    window.removeEventListener('NEOLine.NEO.EVENT.DISCONNECTED', this._disconnected);
                    window.removeEventListener('NEOLine.NEO.EVENT.ACCOUNT_CHANGED', this._disconnected);
                    window.removeEventListener('NEOLine.NEO.EVENT.NETWORK_CHANGED', this._disconnected);
                    this._address = null;
                    this._client = undefined;
                    this._clientCommon = undefined;
                }
                catch (error) {
                    this.emit('error', new WalletDisconnectionError(error === null || error === void 0 ? void 0 : error.message, error));
                }
            }
            this.emit('disconnect');
        });
    }
    invokeRead(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const client = this._client;
            if (!client)
                throw new WalletNotConnectedError();
            try {
                const response = yield client.invokeRead({
                    scriptHash: request.scriptHash,
                    operation: request.operation,
                    args: request.args,
                    signers: request.signers,
                });
                return this._responseToReadResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    invokeReadMulti(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const client = this._client;
            if (!client)
                throw new WalletNotConnectedError();
            try {
                const response = yield client.invokeReadMulti({
                    invokeReadArgs: request.invocations,
                    signers: request.signers,
                });
                return this._responseToReadResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    invoke(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const client = this._client;
            if (!client)
                throw new WalletNotConnectedError();
            try {
                const response = yield client.invoke({
                    scriptHash: request.scriptHash,
                    operation: request.operation,
                    args: request.args,
                    signers: request.signers,
                    fee: request.fee,
                    extraSystemFee: request.extraSystemFee,
                    broadcastOverride: request.broadcastOverride,
                });
                return this._responseToWriteResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    invokeMulti(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const client = this._client;
            if (!client)
                throw new WalletNotConnectedError();
            try {
                const response = yield client.invokeMultiple({
                    invokeArgs: request.invocations,
                    signers: request.signers,
                    fee: request.fee,
                    extraSystemFee: request.extraSystemFee,
                    broadcastOverride: request.broadcastOverride,
                });
                return this._responseToWriteResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    getNetworks() {
        return __awaiter(this, void 0, void 0, function* () {
            const clientCommon = this._clientCommon;
            if (!clientCommon)
                throw new WalletNotConnectedError();
            try {
                const response = yield clientCommon.getNetworks();
                return this._responseToGetNetworksResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    signMessage(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const client = this._client;
            if (!client)
                throw new WalletNotConnectedError();
            try {
                const response = yield client.signMessage({
                    message: request.message,
                });
                return this._responseToSignMessageResult(response);
            }
            catch (error) {
                this.emit('error', error);
                throw error;
            }
        });
    }
    _responseToReadResult(response) {
        // If the state is halt it means that everything went well
        if (response.state === 'HALT') {
            return {
                status: 'success',
                data: {
                    script: response.script,
                    state: response.state,
                    gasConsumed: response.gas_consumed,
                    exception: null,
                    stack: response.stack,
                },
            };
        }
        // Otherwise an error occurred and so return it
        // TODO: message and code where?
        return {
            status: 'error',
            message: undefined,
            code: undefined,
        };
    }
    _responseToWriteResult(response) {
        return {
            status: 'success',
            data: {
                txId: response.txid,
            },
        };
    }
    _responseToGetNetworksResult(response) {
        return {
            status: 'success',
            data: {
                networks: response.networks,
                defaultNetwork: response.defaultNetwork,
            },
        };
    }
    _responseToSignMessageResult(response) {
        return {
            status: 'success',
            data: {
                publicKey: response.publicKey,
                data: response.data,
                salt: response.salt,
                message: response.message,
            },
        };
    }
}
