// @neongd/neo-dapi/dist/cjs/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MethodName = void 0;
var MethodName;
(function (MethodName) {
    MethodName["GetProvider"] = "getProvider";
    MethodName["GetNetworks"] = "getNetworks";
    MethodName["GetAccount"] = "getAccount";
    MethodName["GetBlockCount"] = "getBlockCount";
    MethodName["GetBlock"] = "getBlock";
    MethodName["GetTransaction"] = "getTransaction";
    MethodName["GetApplicationLog"] = "getApplicationLog";
    MethodName["GetStorage"] = "getStorage";
    MethodName["GetNep17Balances"] = "getNep17Balances";
    MethodName["InvokeRead"] = "invokeRead";
    MethodName["InvokeReadMulti"] = "invokeReadMulti";
    MethodName["Invoke"] = "invoke";
    MethodName["InvokeMulti"] = "invokeMulti";
    MethodName["SignMessage"] = "signMessage";
    MethodName["SignMessageWithoutSalt"] = "signMessageWithoutSalt";
    MethodName["SignTransaction"] = "signTransaction";
    MethodName["RelayTransaction"] = "relayTransaction";
})(MethodName = exports.MethodName || (exports.MethodName = {}));
